File: sdk/python/gaussos.py
"""GaussOS Python SDK — a thin, dependency-light client over the REST API.

Uses only the standard library (urllib), so there is nothing to install.

    from gaussos import GaussOS
    gx = GaussOS("http://localhost:8080")
    gx.create_memory("GaussOS uses an embedded SurrealDB backend",
                     tags=["db"], namespace="docs", quality=0.9)
    print(gx.search("surrealdb"))
    print(gx.hybrid_search("embedded database"))
    gx.ingest_fact("GaussOS", "built_by", "Gaussian Technologies")
    print(gx.graph_search(["GaussOS"]))
"""

from __future__ import annotations

import json
import urllib.error
import urllib.request
from typing import Any, Optional


class GaussOSError(RuntimeError):
    pass


class GaussOS:
    """Client for a running GaussOS server."""

    def __init__(self, base_url: str = "http://localhost:8080", timeout: float = 30.0,
                 api_key: Optional[str] = None) -> None:
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self.api_key = api_key

    # ---- low-level ----
    def _request(self, method: str, path: str, body: Optional[dict] = None) -> Any:
        url = f"{self.base_url}{path}"
        data = json.dumps(body).encode() if body is not None else None
        req = urllib.request.Request(url, data=data, method=method)
        req.add_header("Content-Type", "application/json")
        if self.api_key:
            req.add_header("Authorization", f"Bearer {self.api_key}")
        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                raw = resp.read()
                return json.loads(raw) if raw else None
        except urllib.error.HTTPError as e:
            raise GaussOSError(f"HTTP {e.code}: {e.read().decode(errors='replace')}") from e
        except urllib.error.URLError as e:
            raise GaussOSError(f"connection error: {e}") from e

    # ---- health / metrics ----
    def health(self) -> Any:
        return self._request("GET", "/health")

    def metrics(self) -> Any:
        return self._request("GET", "/api/v1/metrics")

    def llm_status(self) -> Any:
        return self._request("GET", "/api/v1/llm/status")

    # ---- memories ----
    def create_memory(self, text: str, *, tags: Optional[list[str]] = None,
                      namespace: str = "default", quality: float = 0.5) -> Any:
        return self._request("POST", "/api/v1/memories", {
            "payload": {"Text": text},
            "tags": tags or [],
            "namespace": namespace,
            "quality_score": quality,
        })

    def search(self, text: str, *, namespace: Optional[str] = None,
               payload_type: Optional[str] = None, min_quality: Optional[float] = None,
               limit: int = 50) -> Any:
        body: dict[str, Any] = {"text": text, "limit": limit}
        if namespace:
            body["namespace"] = namespace
        if payload_type:
            body["payload_type"] = payload_type
        if min_quality is not None:
            body["min_quality"] = min_quality
        return self._request("POST", "/api/v1/memories/search", body)

    def hybrid_search(self, text: str, *, embedding: Optional[list[float]] = None,
                      top_k: int = 10) -> Any:
        body: dict[str, Any] = {"text": text, "hybrid": True, "top_k": top_k}
        if embedding is not None:
            body["embedding"] = embedding
        return self._request("POST", "/api/v1/memories/search", body)

    def compare_retrieval(self, text: str, top_k: int = 10) -> Any:
        """White-box: lexical vs vector vs hybrid with score breakdowns."""
        return self._request("POST", "/api/v1/retrieval/compare", {"text": text, "top_k": top_k})

    def delete_memory(self, memory_id: str) -> Any:
        return self._request("DELETE", f"/api/v1/memories/{memory_id}")

    # ---- bi-temporal facts ----
    def ingest_fact(self, subject: str, predicate: str, obj: str) -> Any:
        return self._request("POST", "/api/v1/facts", {
            "subject": subject, "predicate": predicate, "object": obj,
        })

    def facts_about(self, subject: str) -> Any:
        return self._request("GET", f"/api/v1/facts/{subject}")

    def graph_search(self, seeds: list[str]) -> Any:
        """Multi-hop Personalized-PageRank retrieval over the fact graph."""
        return self._request("POST", "/api/v1/facts/graph-search", {"seeds": seeds})

    def fact_graph(self, at: Optional[str] = None) -> Any:
        path = "/api/v1/facts/graph" + (f"?at={at}" if at else "")
        return self._request("GET", path)

    # ---- admin ----
    def export(self) -> Any:
        return self._request("GET", "/api/v1/admin/export")

    def import_memories(self, memories: list[dict]) -> Any:
        return self._request("POST", "/api/v1/admin/import", {"memories": memories})

    def forget(self, namespace: str, *, delete: bool = False) -> Any:
        return self._request("POST", "/api/v1/admin/forget",
                             {"namespace": namespace, "delete_forgotten": delete})


__all__ = ["GaussOS", "GaussOSError"]

File: sdk/typescript/gaussos.ts
// GaussOS TypeScript SDK — a thin client over the REST API.
// Works in Deno, Node 18+, and browsers (uses the global `fetch`).
//
//   import { GaussOS } from "./gaussos.ts";
//   const gx = new GaussOS("http://localhost:8080");
//   await gx.createMemory("GaussOS uses an embedded SurrealDB backend",
//                         { tags: ["db"], namespace: "docs", quality: 0.9 });
//   console.log(await gx.search("surrealdb"));
//   console.log(await gx.hybridSearch("embedded database"));
//   await gx.ingestFact("GaussOS", "built_by", "Gaussian Technologies");
//   console.log(await gx.graphSearch(["GaussOS"]));

export interface CreateMemoryOpts {
  tags?: string[];
  namespace?: string;
  quality?: number;
}

export interface SearchOpts {
  namespace?: string;
  payloadType?: string;
  minQuality?: number;
  limit?: number;
}

export class GaussOSError extends Error {}

export class GaussOS {
  constructor(
    private baseUrl: string = "http://localhost:8080",
    private apiKey?: string,
  ) {
    this.baseUrl = baseUrl.replace(/\/+$/, "");
  }

  private async request<T = unknown>(
    method: string,
    path: string,
    body?: unknown,
  ): Promise<T> {
    const headers: Record<string, string> = { "Content-Type": "application/json" };
    if (this.apiKey) headers["Authorization"] = `Bearer ${this.apiKey}`;
    const res = await fetch(`${this.baseUrl}${path}`, {
      method,
      headers,
      body: body === undefined ? undefined : JSON.stringify(body),
    });
    if (!res.ok) {
      throw new GaussOSError(`HTTP ${res.status}: ${await res.text()}`);
    }
    const text = await res.text();
    return (text ? JSON.parse(text) : undefined) as T;
  }

  // ---- health / metrics ----
  health(): Promise<unknown> {
    return this.request("GET", "/health");
  }
  metrics(): Promise<unknown> {
    return this.request("GET", "/api/v1/metrics");
  }
  llmStatus(): Promise<unknown> {
    return this.request("GET", "/api/v1/llm/status");
  }

  // ---- memories ----
  createMemory(text: string, opts: CreateMemoryOpts = {}): Promise<unknown> {
    return this.request("POST", "/api/v1/memories", {
      payload: { Text: text },
      tags: opts.tags ?? [],
      namespace: opts.namespace ?? "default",
      quality_score: opts.quality ?? 0.5,
    });
  }

  search(text: string, opts: SearchOpts = {}): Promise<unknown> {
    const body: Record<string, unknown> = { text, limit: opts.limit ?? 50 };
    if (opts.namespace) body.namespace = opts.namespace;
    if (opts.payloadType) body.payload_type = opts.payloadType;
    if (opts.minQuality !== undefined) body.min_quality = opts.minQuality;
    return this.request("POST", "/api/v1/memories/search", body);
  }

  hybridSearch(text: string, embedding?: number[], topK = 10): Promise<unknown> {
    const body: Record<string, unknown> = { text, hybrid: true, top_k: topK };
    if (embedding) body.embedding = embedding;
    return this.request("POST", "/api/v1/memories/search", body);
  }

  /** White-box: lexical vs vector vs hybrid with score breakdowns. */
  compareRetrieval(text: string, topK = 10): Promise<unknown> {
    return this.request("POST", "/api/v1/retrieval/compare", { text, top_k: topK });
  }

  deleteMemory(id: string): Promise<unknown> {
    return this.request("DELETE", `/api/v1/memories/${id}`);
  }

  // ---- bi-temporal facts ----
  ingestFact(subject: string, predicate: string, object: string): Promise<unknown> {
    return this.request("POST", "/api/v1/facts", { subject, predicate, object });
  }
  factsAbout(subject: string): Promise<unknown> {
    return this.request("GET", `/api/v1/facts/${subject}`);
  }
  /** Multi-hop Personalized-PageRank retrieval over the fact graph. */
  graphSearch(seeds: string[]): Promise<unknown> {
    return this.request("POST", "/api/v1/facts/graph-search", { seeds });
  }
  factGraph(at?: string): Promise<unknown> {
    return this.request("GET", "/api/v1/facts/graph" + (at ? `?at=${encodeURIComponent(at)}` : ""));
  }

  // ---- admin ----
  export(): Promise<unknown> {
    return this.request("GET", "/api/v1/admin/export");
  }
  importMemories(memories: unknown[]): Promise<unknown> {
    return this.request("POST", "/api/v1/admin/import", { memories });
  }
  forget(namespace: string, deleteForgotten = false): Promise<unknown> {
    return this.request("POST", "/api/v1/admin/forget", {
      namespace,
      delete_forgotten: deleteForgotten,
    });
  }
}
